src/OrleansRuntime/LogViews/GrainStateWithMetaData.cs

```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime.LogViews
+{
+    /// <summary>
+    /// Used by StorageProviderLogViewAdaptor, which allows a plain storage provider to
+    /// be wrapped by a log view provider, by adding meta data to the grain state.
+    /// </summary>
+    /// <typeparam name="TView">The view state</typeparam>
+    [Serializable]
+    public class GrainStateWithMetaDataAndETag<TView> : IGrainState where TView: class, new()
+    {
+        public GrainStateWithMetaData<TView> StateAndMetaData { get; set; }
+       
+        public string ETag { get; set; }
+
+        object IGrainState.State
+        {
+            get
+            {
+                return StateAndMetaData;
+            }
+            set
+            {
+                StateAndMetaData = (GrainStateWithMetaData<TView>) value;
+            }
+        }
+
+        public GrainStateWithMetaDataAndETag(TView initialview)
+        {
+            StateAndMetaData = new GrainStateWithMetaData<TView>(initialview);
+        }
+        public GrainStateWithMetaDataAndETag()
+        {
+            StateAndMetaData = new GrainStateWithMetaData<TView>();
+        }
+
+        public override string ToString()
+        {
+            return string.Format("v{0} Flags={1} ETag={2} Data={3}", StateAndMetaData.GlobalVersion, StateAndMetaData.WriteVector, ETag, StateAndMetaData.State);
+        }
+    }
+
+
+    [Serializable]
+    public class GrainStateWithMetaData<TView> where TView : class, new()
+    {
+        /// <summary>
+        /// The stored view of the log
+        /// </summary>
+        public TView State { get; set; }
+
+        /// <summary>
+        /// The length of the log
+        /// </summary>
+        public int GlobalVersion { get; set; }
+
+
+        /// <summary>
+        /// Used to avoid duplicate appends.
+        /// A string representing a bit vector, with one bit per replica. 
+        /// Bits are toggled when writing, so that the retry logic can avoid appending an entry twice
+        /// when retrying a failed append. 
+        /// </summary>
+        public string WriteVector { get; set; }
```

Comment:
Ah.  So origin is also a cluster id.  KK.  Still a whole (programmatic config), but it's very small.  Probably safe to ignore.

As for hashset, the hash set was not for speed of access as much as ability to check containment.  You'd mentioned using a dictionary, but if containment check is all that is needed, hashset is simpler.  I don't understand the single access per serialize comment as it's accessed at least twice in the write and once per notification.

I'm not going to push on this too hard, but in general, yes, code maintainability is very important, and writing custom logic to encode multiple fields into a string is not preferable to using a container.  Having said that, the complexity is mostly isolated behind the Toggle and Contains calls, so potential harm is limited.